File: backend/job-application-service/src/main/java/com/jobportal/jobapp/controllers/JobController.java
package com.jobportal.jobapp.controllers;

import com.jobportal.jobapp.dto.JobRequestDto;
import com.jobportal.jobapp.dto.JobResponseDto;
import com.jobportal.jobapp.service.JobService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/jobs")
@Validated
@CrossOrigin(origins = "*")
public class JobController {

    private final JobService jobService;

    public JobController(JobService jobService) {
        this.jobService = jobService;
    }

    @PostMapping
    public ResponseEntity<JobResponseDto> createJob(@Valid @RequestBody JobRequestDto request) {
        return new ResponseEntity<>(jobService.createJob(request), HttpStatus.CREATED);
    }

    @PutMapping("/{id}")
    public ResponseEntity<JobResponseDto> updateJob(@PathVariable Integer id, @Valid @RequestBody JobRequestDto request) {
        return ResponseEntity.ok(jobService.updateJob(id, request));
    }

    @GetMapping("/{id}")
    public ResponseEntity<JobResponseDto> getJobById(@PathVariable Integer id) {
        return ResponseEntity.ok(jobService.getJobById(id));
    }

    @GetMapping("/employer/{empId}")
    public ResponseEntity<List<JobResponseDto>> getJobsByEmployer(@PathVariable Integer empId) {
        return ResponseEntity.ok(jobService.getJobsByEmployer(empId));
    }

    @GetMapping
    public ResponseEntity<List<JobResponseDto>> searchJobs(
            @RequestParam(required = false) String title,
            @RequestParam(required = false) String location,
            @RequestParam(required = false) String type,
            @RequestParam(required = false) String status) {
        if (title == null && location == null && type == null && status == null) {
            return ResponseEntity.ok(jobService.getAllJobs());
        }
        return ResponseEntity.ok(jobService.searchJobs(title, location, type, status));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteJob(@PathVariable Integer id) {
        jobService.deleteJob(id);
        return ResponseEntity.noContent().build();
    }
}

File: backend-dotnet/Services/JobPortal.JobApplicationService/Controllers/JobController.cs
using Microsoft.AspNetCore.Mvc;
using JobPortal.JobApplications.DTOs;
using JobPortal.JobApplications.Interfaces;

namespace JobPortal.JobApplications.Controllers;

[ApiController]
[Route("api/jobs")]
public class JobController : ControllerBase
{
    private readonly IJobService _jobService;

    public JobController(IJobService jobService)
    {
        _jobService = jobService;
    }

    [HttpPost]
    public async Task<IActionResult> CreateJob([FromBody] JobRequestDto request)
    {
        if (!ModelState.IsValid)
        {
            return BadRequest(ModelState);
        }
        var response = await _jobService.CreateJobAsync(request);
        return StatusCode(StatusCodes.Status201Created, response);
    }

    [HttpPut("{id:int}")]
    public async Task<ActionResult<JobResponseDto>> UpdateJob([FromRoute] int id, [FromBody] JobRequestDto request)
    {
        if (!ModelState.IsValid)
        {
            return BadRequest(ModelState);
        }
        var response = await _jobService.UpdateJobAsync(id, request);
        return Ok(response);
    }

    [HttpGet("{id:int}")]
    public async Task<ActionResult<JobResponseDto>> GetJobById([FromRoute] int id)
    {
        var response = await _jobService.GetJobByIdAsync(id);
        return Ok(response);
    }

    [HttpGet("employer/{empId:int}")]
    public async Task<ActionResult<List<JobResponseDto>>> GetJobsByEmployer([FromRoute] int empId)
    {
        var response = await _jobService.GetJobsByEmployerAsync(empId);
        return Ok(response);
    }

    [HttpGet]
    public async Task<ActionResult<List<JobResponseDto>>> SearchJobs(
        [FromQuery] string? title,
        [FromQuery] string? location,
        [FromQuery] string? type,
        [FromQuery] string? status)
    {
        if (title == null && location == null && type == null && status == null)
        {
            var allJobs = await _jobService.GetAllJobsAsync();
            return Ok(allJobs);
        }
        var filteredJobs = await _jobService.SearchJobsAsync(title, location, type, status);
        return Ok(filteredJobs);
    }

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> DeleteJob([FromRoute] int id)
    {
        await _jobService.DeleteJobAsync(id);
        return NoContent();
    }

    [HttpPatch("{jobId:int}/status")]
    public async Task<ActionResult<JobResponseDto>> UpdateJobStatus(
        [FromRoute] int jobId,
        [FromQuery] string status)
    {
        var response = await _jobService.UpdateJobStatusAsync(jobId, status);
        return Ok(response);
    }
}
